File: TEST Inputs/test_input.ts
import WebSocket from 'ws';

const HOST = "10.0.4.20";
const API = `http://${HOST}/api`;
const APP_NAME = "test_input_ts";

const BTN_MAP: Record<number, string> = {
    0: "OK",
    1: "BACK",
    2: "START"
};

async function drawText(text: string) {
    console.log(`Drawing text: ${text}`);
    try {
        await fetch(`${API}/display/draw?application_name=${APP_NAME}`, {
            method: 'DELETE',
            // @ts-ignore - abort signal timeout not strictly needed here for basic test
        });
        
        const payload = {
            application_name: APP_NAME,
            priority: 100,
            elements: [
                {
                    id: "text_1",
                    type: "text",
                    text: text,
                    font: "large",
                    color: "#00FFCCFF",
                    align: "center",
                    x: 36,
                    y: 8,
                    display: "front"
                }
            ]
        };
        await fetch(`${API}/display/draw`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify(payload)
        });
    } catch (e) {
        console.log(`Failed to update display: ${(e as Error).message}`);
    }
}

// -------------------------------------------------------------
// EXACT TS CODE FROM busybar-driver.ts to test the bug!
// -------------------------------------------------------------
function parseVarint(data: Uint8Array, offset: number): { value: number; nextOffset: number } {
  let value = 0;
  let shift = 0;
  let curr = offset;
  while (curr < data.length) {
    const byte = data[curr++];
    value |= (byte & 0x7f) << shift;
    if (!(byte & 0x80)) {
      return { value: value >>> 0, nextOffset: curr };
    }
    shift += 7;
    if (shift > 35) break;
  }
  return { value: 0, nextOffset: data.length };
}

function parseFields(data: Uint8Array): { number: number; wireType: number; value: unknown }[] {
  const fields: { number: number; wireType: number; value: unknown }[] = [];
  let offset = 0;
  while (offset < data.length) {
    const keyRes = parseVarint(data, offset);
    offset = keyRes.nextOffset;
    const key = keyRes.value;
    const number = key >>> 3;
    const wireType = key & 7;

    if (wireType === 0) {
      const valRes = parseVarint(data, offset);
      offset = valRes.nextOffset;
      fields.push({ number, wireType, value: valRes.value });
    } else if (wireType === 1) {
      const sub = data.subarray(offset, offset + 8);
      offset += 8;
      fields.push({ number, wireType, value: sub });
    } else if (wireType === 2) {
      const lenRes = parseVarint(data, offset);
      offset = lenRes.nextOffset;
      const len = lenRes.value;
      const sub = data.subarray(offset, offset + len);
      offset += len;
      fields.push({ number, wireType, value: sub });
    } else if (wireType === 5) {
      const sub = data.subarray(offset, offset + 4);
      offset += 4;
      fields.push({ number, wireType, value: sub });
    } else {
      break;
    }
  }
  return fields;
}

function zigzagDecode(val: number): number {
  return (val >> 1) ^ -(val & 1);
}

function processInputEventBytes(value: Uint8Array) {
    const fields = parseFields(value);
    for (const f of fields) {
        if (f.number === 1 && f.value instanceof Uint8Array) {
            let button = 0;
            let action = 0;
            const subFields = parseFields(f.value);
            for (const sf of subFields) {
                if (sf.number === 1 && typeof sf.value === 'number') button = sf.value;
                if (sf.number === 2 && typeof sf.value === 'number') action = sf.value;
            }
            return { kind: 'button', button, action };
        }
        if (f.number === 2 && f.value instanceof Uint8Array) {
            let pos = 0;
            const subFields = parseFields(f.value);
            for (const sf of subFields) {
                if (sf.number === 1 && typeof sf.value === 'number') pos = sf.value;
            }
            return { kind: 'switch', pos };
        }
        if (f.number === 3 && f.value instanceof Uint8Array) {
            let delta = 0;
            const subFields = parseFields(f.value);
            for (const sf of subFields) {
                if (sf.number === 1 && typeof sf.value === 'number') delta = zigzagDecode(sf.value);
            }
            return { kind: 'encoder', delta };
        }
    }
    return null;
}

function eventsFromState(data: Uint8Array) {
    const events: any[] = [];
    const rootFields = parseFields(data);
    for (const rf of rootFields) {
        if (rf.number === 2 && rf.value instanceof Uint8Array) {
            const updateFields = parseFields(rf.value);
            for (const uf of updateFields) {
                if (uf.number === 11 && uf.value instanceof Uint8Array) {
                    const res = processInputEventBytes(uf.value);
                    if (res) events.push(res);
                }
            }
        } else if (rf.number === 11 && rf.value instanceof Uint8Array) {
            const res = processInputEventBytes(rf.value);
            if (res) events.push(res);
        }
    }
    return events;
}

async function main() {
    const wsUrl = `ws://${HOST}/api/status/ws`;
    console.log(`Connecting to ${wsUrl}...`);
    
    await drawText("WAITING (TS)");

    function connect() {
        const socket = new WebSocket(wsUrl);
        
        socket.on('open', () => {
            console.log("Connected! Waiting for physical button inputs...");
            socket.send(JSON.stringify({ enable: true }));
        });

        socket.on('message', (rawData) => {
            let data: Uint8Array;
            if (Buffer.isBuffer(rawData)) {
                data = new Uint8Array(rawData);
            } else if (rawData instanceof ArrayBuffer) {
                data = new Uint8Array(rawData);
            } else if (Array.isArray(rawData)) {
                data = new Uint8Array(Buffer.concat(rawData));
            } else {
                return; // String or unhandled
            }

            try {
                const events = eventsFromState(data);
                for (const event of events) {
                    if (event.kind === 'button') {
                        if (event.action === 1) { // Release
                            const btnName = BTN_MAP[event.button] || `BTN ${event.button}`;
                            console.log(`Detected: ${btnName}`);
                            drawText(btnName);
                        }
                    } else if (event.kind === 'switch') {
                        const mode = event.pos === 3 ? "APPS" : `MODE ${event.pos}`;
                        console.log(`Detected Switch: ${mode}`);
                        drawText(mode);
                    } else if (event.kind === 'encoder') {
                        const dirStr = event.delta > 0 ? "RIGHT" : "LEFT";
                        console.log(`Detected Encoder: ${dirStr}`);
                        drawText(dirStr);
                    }
                }
            } catch (e) {
                console.error("Parse error:", (e as Error).message);
            }
        });

        socket.on('error', (err) => {
            console.log(`Connection error: ${err.message}`);
        });

        socket.on('close', () => {
            console.log("Connection lost. Retrying in 2s...");
            setTimeout(connect, 2000);
        });
    }

    connect();
}

main().catch(err => {
    console.error("Fatal error:", err);
});

File: Documentation/hardware-samples/input-websocket/test_input.py
#!/usr/bin/env python3
import time
import requests
from websocket import create_connection, WebSocketTimeoutException
from collections.abc import Iterator

HOST = "10.0.4.20"
API = f"http://{HOST}/api"
APP_NAME = "test_input"

BTN_MAP = {
    0: "OK",
    1: "BACK",
    2: "START"
}

def draw_text(text: str) -> None:
    print(f"Drawing text: {text}")
    try:
        # Clear previous drawing
        requests.delete(f"{API}/display/draw", params={"application_name": APP_NAME}, timeout=2)
        
        # Draw new text
        payload = {
            "application_name": APP_NAME,
            "priority": 100,
            "elements": [
                {
                    "id": "text_1",
                    "type": "text",
                    "text": text,
                    "font": "large",
                    "color": "#00FFCCFF",
                    "align": "center",
                    "x": 36,
                    "y": 8,
                    "display": "front"
                }
            ]
        }
        requests.post(f"{API}/display/draw", json=payload, timeout=2)
    except Exception as e:
        print(f"Failed to update display: {e}")

def _read_varint(data: bytes, offset: int) -> tuple[int, int]:
    value = 0
    shift = 0
    while offset < len(data):
        byte = data[offset]
        offset += 1
        value |= (byte & 0x7F) << shift
        if not byte & 0x80:
            return value, offset
        shift += 7
        if shift > 70:
            break
    raise ValueError("Invalid protobuf varint")

def _fields(data: bytes) -> Iterator[tuple[int, int, int | bytes]]:
    offset = 0
    while offset < len(data):
        key, offset = _read_varint(data, offset)
        number, wire_type = key >> 3, key & 7
        if wire_type == 0:
            value, offset = _read_varint(data, offset)
        elif wire_type == 1:
            value = data[offset : offset + 8]
            offset += 8
        elif wire_type == 2:
            length, offset = _read_varint(data, offset)
            value = data[offset : offset + length]
            offset += length
        elif wire_type == 5:
            value = data[offset : offset + 4]
            offset += 4
        else:
            raise ValueError(f"Unsupported protobuf wire type: {wire_type}")
        yield number, wire_type, value

def _zigzag(value: int) -> int:
    return (value >> 1) ^ -(value & 1)

def _input_event(data: bytes) -> tuple | None:
    for number, wire_type, value in _fields(data):
        if wire_type != 2 or not isinstance(value, bytes):
            continue
        if number == 1:
            button = action = 0
            for field, _, item in _fields(value):
                if field == 1 and isinstance(item, int):
                    button = item
                elif field == 2 and isinstance(item, int):
                    action = item
            return "button", button, action
        if number == 2:
            position = 0
            for field, _, item in _fields(value):
                if field == 1 and isinstance(item, int):
                    position = item
            return "switch", position
        if number == 3:
            delta = 0
            for field, _, item in _fields(value):
                if field == 1 and isinstance(item, int):
                    delta = _zigzag(item)
            return "encoder", delta
    return None

def events_from_state(data: bytes) -> Iterator[tuple]:
    for number, wire_type, update in _fields(data):
        if number != 2 or wire_type != 2 or not isinstance(update, bytes):
            continue
        for field, update_wire_type, value in _fields(update):
            if field == 11 and update_wire_type == 2 and isinstance(value, bytes):
                event = _input_event(value)
                if event:
                    yield event

def main():
    ws_url = f"ws://{HOST}/api/status/ws"
    print(f"Connecting to {ws_url}...")
    
    draw_text("WAITING")
    
    while True:
        socket = None
        try:
            socket = create_connection(ws_url, timeout=5, enable_multithread=True)
            socket.settimeout(1)
            socket.send('{"enable":true}')
            print("Connected! Waiting for physical button inputs...")
            
            while True:
                try:
                    opcode, data = socket.recv_data(control_frame=False)
                except WebSocketTimeoutException:
                    continue
                if opcode != 2 or not data:
                    continue
                
                for event in events_from_state(data):
                    kind = event[0]
                    if kind == "button":
                        button_id = int(event[1])
                        action = int(event[2])
                        if action == 1:  # Release
                            btn_name = BTN_MAP.get(button_id, f"BTN {button_id}")
                            print(f"Detected: {btn_name}")
                            draw_text(btn_name)
                    elif kind == "switch":
                        pos = int(event[1])
                        mode = "APPS" if pos == 3 else f"MODE {pos}"
                        print(f"Detected Switch: {mode}")
                        draw_text(mode)
                    elif kind == "encoder":
                        delta = int(event[1])
                        dir_str = "RIGHT" if delta > 0 else "LEFT"
                        print(f"Detected Encoder: {dir_str}")
                        draw_text(dir_str)
        except Exception as error:
            print(f"Connection lost: {error}. Retrying in 2s...")
            time.sleep(2)
        finally:
            if socket:
                try:
                    socket.close()
                except Exception:
                    pass

if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        print("\nExiting test script.")
        try:
            requests.delete(f"{API}/display/draw", params={"application_name": APP_NAME}, timeout=2)
        except Exception:
            pass
